libraries/AP_HAL_ChibiOS/hwdef/ARK_FPV/hwdef.dat

```diff
@@ -0,0 +1,219 @@
+# Processed by chibios_hwdef.py
+
+include common.inc
+
+# ChibiOS system timer
+STM32_ST_USE_TIMER 2
+
+# Bootloader takes first sector
+FLASH_RESERVE_START_KB 128
+
+# Use last 2 pages for flash storage
+# H743 has 16 pages of 128k each
+define HAL_STORAGE_SIZE 32768
+STORAGE_FLASH_PAGE 14
+
+# To be compatible with the px4 bootloader we need
+# to use a different RAM_MAP
+env USE_ALT_RAM_MAP 1
+
+env OPTIMIZE -Os
+
+SERIAL_ORDER OTG1 UART7 UART5 USART1 USART2 UART4 USART6 OTG2
+
+define DEFAULT_SERIAL6_PROTOCOL SerialProtocol_RCIN
+define DEFAULT_SERIAL7_PROTOCOL SerialProtocol_SLCAN
+
+# Debug console
+STDOUT_SERIAL SD3
+STDOUT_BAUDRATE 57600
+
+# Default to all pins low to avoid ESD issues
+DEFAULTGPIO OUTPUT LOW PULLDOWN
+
+# Telem1 - Telem
+PE8  UART7_TX  UART7
+PF6  UART7_RX  UART7
+PF8  UART7_RTS UART7
+PE10 UART7_CTS UART7
+define DEFAULT_SERIAL1_PROTOCOL SerialProtocol_MAVLink2
+define DEFAULT_SERIAL1_BAUD 115
+
+# Telem2 - VTX (DJI Air Unit)
+PC12 UART5_TX UART5
+PD2  UART5_RX UART5
+define DEFAULT_SERIAL2_PROTOCOL SerialProtocol_MSP_DisplayPort
+define DEFAULT_SERIAL2_BAUD 115
+
+# GPS1
+PB6 USART1_TX USART1
+PB7 USART1_RX USART1
+define DEFAULT_SERIAL3_PROTOCOL SerialProtocol_GPS
+
+# Telem3 - VTX (DJI Air Unit, RX only)
+PA3 USART2_RX USART2
+PD5 USART2_TX USART2
+define DEFAULT_SERIAL4_PROTOCOL SerialProtocol_None # TODO: DJI HDL?
+
+# ESC telem RX only
+PH13 UART4_TX UART4
+PH14 UART4_RX UART4
+define DEFAULT_SERIAL5_PROTOCOL SerialProtocol_ESCTelemetry
+define DEFAULT_SERIAL5_BAUD 115
+
+# RC Input
+PC6 USART6_TX USART6
+PC7 USART6_RX USART6
+
+# ADCs
+PA0 SCALED1_V3V3 ADC1 SCALE(2)
+PB1 VDD_5V_SENS ADC1 SCALE(2)
+# TODO: 12V monitor
+# PA4 SCALED2_V3V3 ADC1 SCALE(7.66)
+
+PB0 BATT_VOLTAGE_SENS ADC1 SCALE(1)
+PC2 BATT_CURRENT_SENS ADC1 SCALE(1)
+
+# Analog battery input scaling
+# See STM32H743xx.py
+define HAL_BATT_VOLT_PIN 9
+define HAL_BATT_CURR_PIN 12
+define HAL_BATT_VOLT_SCALE 21
+define HAL_BATT_CURR_SCALE 120
+define HAL_BATT_MONITOR_DEFAULT 4
+
+# ESC Telem on UART4
+define HAL_SERIAL_ESC_COMM_ENABLED 1
+define HAL_WITH_ESC_TELEM 1
+define DEFAULT_SERIAL4_PROTOCOL SerialProtocol_ESCTelemetry
+
+# SPI1 - IIM42653 IMU
+PA5 SPI1_SCK SPI1
+PB5 SPI1_MOSI SPI1
+PG9 SPI1_MISO SPI1
+PI9 IMU1_CS CS
+PF2 IMU1_DRDY INPUT
+
+# SPI6 - external (SPI OSD or IMU)
+PB3  SPI6_SCK SPI6
+PA6  SPI6_MISO SPI6
+PG14 SPI6_MOSI SPI6
+PI10 EXT1_CS CS
+PD11 DRDY_ADIS16507 INPUT GPIO(93)
+
+# GPIO(93) for data ready on ADIS16507
+# define ADIS_DRDY_PIN 93
+
+# Motors, see STM32H743xx.py
+PI0  TIM5_CH4 TIM5 PWM(1) GPIO(50) BIDIR
+PH12 TIM5_CH3 TIM5 PWM(2) GPIO(51) BIDIR
+PH11 TIM5_CH2 TIM5 PWM(3) GPIO(52) BIDIR
+PH10 TIM5_CH1 TIM5 PWM(4) GPIO(53) BIDIR
+PI5  TIM8_CH1 TIM8 PWM(5) GPIO(54) BIDIR
+PI6  TIM8_CH2 TIM8 PWM(6) GPIO(55) BIDIR
+PI7  TIM8_CH3 TIM8 PWM(7) GPIO(56) BIDIR
+PI2  TIM8_CH4 TIM8 PWM(8) GPIO(57) BIDIR
```

Comment:
remove bidir

Reply:
how did you determine the correct ones to select for BIDIR? I was reading through some forum posts where you explained it, I thought as long as they didn't share DMA it was okay. I'm also surprised the group is split across two timers. I recently worked on bidir dshot in PX4 so I would love to hear the technical reasons and understand how it works in ardupilot.

Reply:
should I set any new default params for bidir dshot? Or better to let the user configure these?

MOT_PWM_TYPE
SERVO_BLH_BDMASK
SERVO_BLH_OTYPE
SERVO_DSHOT_ESC
SERVO_DSHOT_RATE

Reply:
Unless you know what ESC the user is going to use I would leave these alone

Reply:
Just tested this and it works
```
PI0  TIM5_CH4 TIM5 PWM(1) GPIO(50) BIDIR
PH12 TIM5_CH3 TIM5 PWM(2) GPIO(51) BIDIR
PH11 TIM5_CH2 TIM5 PWM(3) GPIO(52) BIDIR
PH10 TIM5_CH1 TIM5 PWM(4) GPIO(53) BIDIR
PI5  TIM8_CH1 TIM8 PWM(5) GPIO(54)
PI6  TIM8_CH2 TIM8 PWM(6) GPIO(55)
PI7  TIM8_CH3 TIM8 PWM(7) GPIO(56)
PI2  TIM8_CH4 TIM8 PWM(8) GPIO(57)
PD12 TIM4_CH1 TIM4 PWM(9) GPIO(58)
```

Reply:
We rotate the inbound DMA channel between CH1-2 and CH3-4, this is because the IC is organized in this way. I suspect that what yo uhave above will simply mean that the first channel in the IC group gets used. Removing one makes it clear which one is being used because sometimes it matters.

Reply:
> We rotate the inbound DMA channel between CH1-2 and CH3-4

Are you saying that the dshot RX frame is being read round robin? I am looking at the hwdef.dat and this is how it appears since they all are configured use the same DMA stream when I group them all on a single timer. So the eRPM read rate would be 1/4th.
```
#define STM32_TIM_TIM5_CH1_DMA_STREAM  STM32_DMA_STREAM_ID(2, 2)
#define STM32_TIM_TIM5_CH2_DMA_STREAM  STM32_DMA_STREAM_ID(2, 2)
#define STM32_TIM_TIM5_CH3_DMA_STREAM  STM32_DMA_STREAM_ID(2, 2)
#define STM32_TIM_TIM5_CH4_DMA_STREAM  STM32_DMA_STREAM_ID(2, 2)
```

Reply:
For each dshot pulse that goes out using the timer _UP channel only one is read on the return. So yes in this case the rate is 1/4 of the output rate. This is the main reason that the SERVO_DSHOT_RATE parameter exists.